File: src/main/scala/Race/common/CarState.scala
package Race.common

import java.awt.Polygon

object CarState {

  def simpleState(p: Point, rad: Double): CarState = SimpleStateImpl(p, rad)

  private def rotatedCar(origin: Point, rad: Double): Polygon = {
    val points = Array(Point(-5, 10), Point(5, 10), Point(5, -10), Point(-5, -10))
      .map(_.rotate(rad))
      .map(_ + origin)

    new Polygon(points.map(_.x), points.map(_.y), points.length)
  }

  def standardLidarOrigins: List[Point] =
    List(
      Point(0, -1000),
      Point(-500, -1000),
      Point(500, -1000),
      Point(-1000, -700),
      Point(1000, -700)
    )

  case class SimpleStateImpl(centerOfMass: Point, direction: Double) extends CarState {
    override def position: Polygon = rotatedCar(centerOfMass, direction)
    override def lidar: List[(Point, Point)] = {
      standardLidarOrigins
        .map(_.rotate(direction))
        .map(p => (centerOfMass, p + centerOfMass))
    }

    override def speed: Double = 0
  }
}


trait CarState {
  def centerOfMass: Point
  def position: Polygon
  def lidar: List[(Point, Point)]
  def speed: Double
  def direction: Double
}

File: PyCar/common/car_state.py
from PyQt5.QtGui import QPolygon

from common.point import Point


class CarState:
    def __init__(self, center_of_mass: Point, direction: float, momentum: (float, float)):
        self.center_of_mass = center_of_mass
        self.momentum = momentum
        self.direction = direction

    def position(self) -> QPolygon:
        points = [Point(-5, 10), Point(5, 10), Point(5, -10), Point(-5, -10)]
        points = map(lambda p: p.rotate(self.direction) + self.center_of_mass, points)
        return QPolygon(points)
